src/protocol/commands/_qXfer_exec_file.rs

```diff
@@ -3,27 +3,32 @@ use super::prelude::*;
 use crate::common::Pid;
 
 #[derive(Debug)]
-pub struct qXferExecFileRead {
+pub struct qXferExecFileRead<'a> {
     pub pid: Option<Pid>,
-    pub offset: usize,
-    pub len: usize,
+    pub offset: &'a [u8],
+    pub length: &'a [u8],
```

Comment:
I'm pretty sure these can stay as `usize` values, since qXfer packets are reading data from the remote gdbserver, vs. the remote target being debugged, so it's appropriate to use the host's `usize` here.

This should also make it possible to use the _entire_ PacketBuf in this case, as you won't have any dangling references

Reply:
Are you sure? But in theory the data can be read from a file, and the file may be very large.

Reply:
Yes, indeed, but the file that would be read (with qXfer packets) would most likely reside on the "Host" rather than the "Guest/Inferior/Target" being debugged, and as such, the integral you'd be using to read/write from the file would be the host's native `usize` (e.g: via std::io::Read/Write).

Reply:
> the file that would be read would most likely reside on the "Host"

It's really common to get exec file by reading `/proc/self/exe`, which locates on the "Target"

Reply:
I think the confusion here is where we each draw the line between "Host" and "Target"

A typical `gdbserver` instance would probably _not_ use `/proc/self`, as it would be debugging inferior processes rather than itself. In such a case, `usize` is correct, as `proc/<pid>/exe` would be read in the context of the Host fs. In the case you're describing, which I presume to be an "in process" debugger, the target and the host are one in the same, but if that's the case, then `usize` is _still_ appropriate, as that would be the correct "usize" on the target 😄

Reply:
`gdbserver` is the "Target", right? `/proc/pid/exe` is also on the Target?
If we use `usize` here, why we not use `usize` in `vFile:pread`?

Reply:
No, `gdbserver` is the "Host". The inferior process / the thing that is being debugged is the "Target".

`vFile:pread` reads data from a inferior's filesystem, and as such, should use the inferior (i.e: the target's) usize. `qXfer` packets read data from the host's filesystem, and as such, should use the host's usize.

Reply:
What, `gdbserver` and the process being debugged can be not in a same place? Could you give me an example?
I know `gdb` and `gdbserver` can be in a different place, but `gdbserver` ( or the project integrate `gdbstub` ) should always be in the same place with the process being debugged?

Reply:
For a while now, it's been possible to debug multiple processes of different architectures via a single gdbserver process (here is a link to the [very old PR](https://sourceware.org/bugzilla/show_bug.cgi?id=14823) that introduced this functionality). e.g: a 64 bit gdbserver process can debug a 32-bit process on a machine that supports running mixed 32 bit and 64 bit code. similarly, the new M1 based macs support debugging x86 and ARM code via a single gdbserver as well.

Note that at the moment, `gdbstub` doesn't actually have the right infrastructure in place to support simultaneous multi-arch debugging, but it is something on the long term roadmap.

Reply:
If I'm not mistaken, the `gdb` client, `gdbserver` and the process being debugged  can be in three different places with three different `usize`?

Reply:
yep, that's right! isn't that cool 😄

Reply:
Then shouldn't `vFile:pread` also use `usize`?
Do you want to say that the "Host" in "Host I/O" is not the same with the "Host" we are talking about?

Reply:
You know... the more I read through the docs, the more murky this host / target distinction becomes. 

The whole arch-dependent usize thing was pretty important when it came to implementing memory/register access parts of the protocol, but I'm starting to think it might not matter [as much] for [most] qXfer and vFile packets.

While it's always nice to hit the "perfect" API on the first try, I suspect that until we have `gdbstub` deployed in more projects, and we start getting real-world feedback on the APIs, we won't know whether we've made the right move.

At this point, I think I'm leaning towards "throwing in the towel" and just using regular 'ol `usize` in vFile as well. IIRC you're already planning on sending a follow up PR that plumbs PacketBuf through to those APIs, so you might as well include this change along with those as well.

* * *

Also, quick meta comment: as it should be extremely clear by now, I'm also just figuring things out about the GDB RSP as I go, and thusfar, I've basically just "gone with my gut" on a lot of design decisions. It's been nice to have someone to bounce ideas off of, so, thank you for the good conversation and for raising interesting points!

Reply:
Can a 32-bit operating system read files larger than 4G? If it can, the offset should always be 64-bit.
I think of it because:
https://github.com/daniel5151/gdbstub/blob/9227dfd0b78db5b20859d13c890f6f47e597029e/examples/armv4t/gdb/host_io.rs#L144

Reply:
Good catch!

It seems the way this is handled in Rust is by having the seek API take a `u64` offset, while the read/write APIs still use `usize`. i.e: you can only read a contiguous `usize` bytes at a time, but the underlying file may be larger than `usize` in length.

So yes, it seems we should be using `u64` for the `offset` type, the same way Rust does it. The _perfect_ solution would actually be to add a new type param to `Arch` that allows specifying what type to use, but plumbing that through would be a real PITA. Maybe I'll give it a shot one day

My mind keeps wandering to the degenerate case of implementing `gdbstub` on a 8-bit AVR, and being forced to handle `u64` file offsets, but honestly, I probably shouldn't worry about that 😅

Reply:
So do we need to set the type of `offset` to `u64` in this PR?

Reply:
Yes, I think we should make `offset: u64` in this PR, and also change the relevant `vFile` APIs in a follow up PR.